### ISSUE TITLE
Default map at /map; move current metadata to /styles and {collectionId}

### ISSUE THREAD

**Issue by jerstlouis** [opengeospatial/ogcapi-maps#56]
2020-09-03 SWG - I proposed today that since the `/map` and `/styles` resources are eerily similar (case in point: http://maps.ecere.com/ogcapi/collections/Daraa/styles vs. http://maps.ecere.com/ogcapi/collections/Daraa/map -- both consist primarily in a `styles` property with a list of styles), it would make a lot of sense to merge those resources (API could provide links to either or both style sheets and maps for each styles). This would allow to have a default map at `/map` directly, in addition to optional maps for different styles at `/map/{styleId}`.

Any other necessary information about maps which is not already covered by the collection resource at `.../{collectionId}` or the landing page at `{datasetAPI}/` (in the case of maps for whole dataset) could be added there as an extension.

Potentially this extra information to provide is minimal, as extent and CRS information should already be available (though concerns were previously raised about a different set of CRS being usable for maps resources -- this may require some discussion).

A default map directly at `/map` improves the symmetry with the Features and Coverages APIs, and makes for a less awkward URI (no dummy `default` style) when you are using the Maps API as a mechanism for retrieving either simple imagery without a style, or retrieving a map for a style you are providing in a POST request, or retrieving the results of a process generating an image (e.g. Modular OGC API Workflows approach: https://app.swaggerhub.com/apis-docs/jerstlouis/MOAW/MOAW-0.1 could remove those `/default`). It also solves the diffculty of having to know what the default style is called ( https://github.com/opengeospatial/OGC-API-Maps/issues/54 ).

**Comment by joanma747** [comment]
At the beginning of the maps and tiles discussions we add some extra properties for maps in the .../{collectionId} but in a TC plenary  (in Toulouse I remember correctly) some OGC API feature people preferred to remove them and move them to a"maps only" place. This is how we end up with .../{collectionId}/maps "maps description document".
There is a value on having .../{collectionId}/maps: we do not have to extend the .../{collectionId} in the OGC API maps. We only use one of the links to point to .../{collectionId}/maps. 
I do not remember the rational to separate the metadata. If there is a rational please reproduce it here. It the rational have been OBE (it seems that coverages ans injecting stuff in .../{collectionId} then, ther eis no reason to go back to the original design.

Please thing about this carefully as one of this changes represents hours in the draft writing and swagger hub etc, and more probability of mistakes.

**Comment by pomakis** [comment]
So let me get this straight.  I believe your proposal is more or less to move the collections/{collectionId}/map metadata endpoint (as Ecere and CubeWerx currently implement it) to collections/{collectionId}/styles in order to free up collections/{collectionId}/map for use as a default-style map shortcut endpoint.

I have no particular problem with doing that.  However, a default-style map shortcut endpoint has some shortcomings that need to be explicitly stated (or somehow resolved).  Specifically, there's no safe way to build on it like with the canonical "map/{styleId}" endpoints.  For example, if one wants to perform a feature-info query on a map of a particular style, one can simply append "/info?..." to the "map/{styleId}" URL.  But if one wants to perform a feature-info query on a map of the default style (without explicitly looking up and providing the actual styleId of the default style), it doesn't seem safe to allow "map/info".  We'd have to forbid "tiles", "legend", "info", etc., from being used as style IDs.  That doesn't seem reasonable.  So the default-style map shortcut endpoint would really only be useable in situations where feature-info queries, etc., are not needed.  I.e., there can be no optional sub-path elements.  And from what I can see, there won't be any href links to the default-style map shortcut endpoint.  It'll be knowable through the specification and OpenAPI document only.

Also, the advertised server constraints like "maxWidth" and "maxHeight" would seem awkward at a collections/{collectionId}/styles endpoint.  But there has already been talk of providing a more generic mechanism for advertising server constraints anyways, so perhaps that's not a problem.

**Comment by jerstlouis** [comment]
@pomakis Good point about the /{styleId} potentially clasing with tiles, info etc.

I wonder if we could perhaps have `/map` appended to `/styles/{styleId}` instead? e.g. `{collectionId}/styles/{styleId}/map`

Seems like that would solve that particular issue. Another way (which some might find awkward, and might not fly too well with OpenAPI, is reserved words that could not be valid styles... tiles, info...).

About `maxWidth` and `maxHeight` yes that was also my idea to have server constrains all grouped somewhere.

There remains the question about CRSes which could differ for a map rather than for the raw data, but maybe again this is a server-wide setting which could be advertised somewhere else?

**Comment by pomakis** [comment]
@jerstlouis wrote:
> I wonder if we could perhaps have /map appended to /styles/{styleId} instead? e.g. {collectionId}/styles/{styleId}/map

A {collectionId}/styles/{styleId}/map endpoint would still require the use of the "default" keyword.  But I guess it's not as much of a conceptual problem in this case.  So you'd have:

[collections/{collectionId}/]styles
  - returns a list of styles, each with a bunch of links, similar to the existing implementations at "http://maps.ecere.com/ogcapi/collections/Daraa/map?f=json" or "https://test.cubewerx.com/cubewerx/cubeserv/demo/ogcapi/Daraa/collections/TransportationGroundCrv/map?f=json&pretty=true"

[collections/{collectionId}/]styles/{styleId}
  - returns the specified style, replete with all of its links
  - the "default" keyword requests the default style

[collections/{collectionId}/]styles/{styleId}/map
  - returns a map of the specified style
  - accepts a bunch of optional query parameters
  - the "default" keyword requests a map of the default style

[collections/{collectionId}/]styles/{styleId}/map/info/...
  - performs a feature-info query on a map of the specified style
  - accepts the same optional query parameters as {styleId}/map
  - the "default" keyword performs a feature-info query on a map of the default style

[collections/{collectionId}/]styles/{styleId}/legend
  - returns a legend graphic for the specified style
  - the "default" keyword requests a map of the default style
  - a scale parameter can be used to request a legend graphic for a particular scale

[collections/{collectionId}/]styles/{styleId}/tiles
  - returns a metadata document describing the tiles that are available in the specified style
  - the "default" keyword requests a tiles for the default style

Hmmm, I think I like it.

**Comment by jerstlouis** [comment]
@pomakis my idea was to have the `/styles/{styleId}/map` as optional resources in addition to the default `/map`.

whether `/styles/default` is required or not (and corresponds to `/map`) can be debated.

**Comment by pomakis** [comment]
> @pomakis my idea was to have the /styles/{styleId}/map as optional resources in addition to the default /map.

That just seems like pointless redundancy.

**Comment by jerstlouis** [comment]
Well I was seeing `/map` as the default map for the collection, and `/styles/{styleId}/map` as a map for a particular style.

Makes a lot of sense to me, especially if you consider that `/map` might be the API by which you simply share static imagery in a single default style, where `/styles` would really be awkward.

**Comment by pomakis** [comment]
Okay, so you're suggesting this:

map
  - returns a map
  - accepts a bunch of optional query parameters

map/info/...
  - performs a feature-info query on a map
  - accepts the same optional query parameters as {styleId}/map

map/legend
  - returns a legend graphic for a map
  - a scale parameter can be used to request a legend graphic for a particular scale

map/tiles
  - returns a metadata document describing the map tiles that are available 

And these map endpoints can exist in at least the following locations in the OGC API endpoint hierarchy:

{landingPage}
  - for requesting a map of the whole dataset (or the specified subset) in the default style

{landingPage}/styles/{styleId}
  - for requesting a map of the whole dataset (or the specified subset) in a specific style

{landingPage}/collections/{collectionId}
  - for requesting a map of a specific collection in the default style

{landingPage}/collections/{collectionId}/styles/{styleId}
  - for requesting a map of a specific collection in a specific style

And there's *no* "default" keyword.

Yikes, that's a total of 4x4 = 16 different endpoints, making for a pretty hefty OpenAPI document.  But I like the orthogonality of it all.

**Comment by jerstlouis** [comment]
@pomakis correct. This orthogonality is what modular building blocks are all about in my opinion :)

(e.g. for modular workflows we allow everything that goes after `/collections/{collectionId}` after `/processes/{processId}`)

**Comment by pomakis** [comment]
The more I think about these endpoints, the more they grow on me.  So much so that I'm really hoping that they get accepted.

I've gone ahead and implemented these endpoints as an experiment to see how it pans out.  And I like it!  I think this endpoint organization makes a lot more sense than what we currently have, and really hope that the "OGC API - Maps" and "OGC API - Styles" specifications-in-progress can be adjusted accordingly.  (I don't think "OGC API - Tiles" is affected, since it already orthogonally hangs /tiles off of anything tileable.)

These experimental endpoint are available at the following server:
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa

For example:
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/map
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/map?style=Topographic
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/styles?f=json&pretty=true
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/styles/Topographic/map
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/map
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/map?style=Topographic
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/map/legend?transparent=false
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/map/tiles?f=json&pretty=true
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/styles?f=json&pretty=true
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/styles/Topographic?f=json&pretty=true
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/styles/Topographic/map
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/styles/Topographic/map/legend?transparent=false
    https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/styles/Topographic/map/tiles?f=json&pretty=true

This is all hacked in at the moment, and not quite complete.  E.g., the {landingPage}/styles/{styleId) endpoint isn't properly implemented yet, and neither are the HTML versions of many of these endpoints.  And the OpenAPI document doesn't report these endoints.

One thing you may notice is that we accept an optional "styleId" parameter on the map endpoints that don't hang off a {styleId}.  I'm not necessarily proposing this parameter as part of the standard, but just figured I'd point out the possibility of such a parameter.

**Comment by jeffharrison** [comment]
@pomakis I don't know about others, but I'd say the "OGC API - Styles" specification-in-progress should definitely take a look at this during an upcoming SWG meeting.

I'm still clicking through all the links... great examples!

Thanks and Regards,
Jeff

**Comment by jerstlouis** [comment]
@pomakis I love all of these :)

**Comment by jerstlouis** [comment]
For the style resource in JSON though, do we really need another metadata resource, since the style sheets links are already included at the level up?

https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/styles/Topographic?f=json&pretty=true

Our client currently would have expected this to be a style sheet already in MBGL style since it uses JSON, e.g.:

https://maps.ecere.com/ogcapi/collections/Daraa:TransportationGroundCrv/styles/topographic?f=json

The default or html current is not implemented (possibly this would be a preview of the style with link to different encoding):

https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/styles/Topographic

**Comment by pomakis** [comment]
I took a cue from the collections and collections/{collectionId} endpoints in that the /collections endpoint returns an array of all of the available collections (as full collection objects), while the collections/{collectionId} endpoint returns just the specified one.  I know there's been talk of the collection objects reported by the collections endpoint being a subset of what gets reported by the collections/{collectionId} endpoint, but a) I think that'd be awkward, and b) it's still (at least a superset or subset of) the same object either way.

I like this organization because it allows clients (or people for that matter) to pass around references to a specific collections/{collectionId}/styles/{styleId} endpoint, which in theory provides the receiver with everything there is to know about that style for that layer, including how to make map, tile, and legend requests, etc., for that layer.  Similarly, the resource at {landingPage}/styles/{styleId} should provide everything there is to know about that style for the whole data set, including how to make map, tile, and legend requests, etc.  I see it looking exactly like the collections/{collectionId}/styles/{styleId} resource, except for an extra "compatibleCollections" field indicating which collections the style is relevant for (and therefore what can be listed in the optional "geodata" parameter at the map endpoint under it).

I also think that it makes conceptual sense for the resource at .../styles/{styleId} to provide links to the resources under it, e.g., .../styles/{styleId}/map, etc.  Having the various {styleId} endpoints return SLD or MapBox style definitions when requested to (via the Accept header or f parameter) is fine, but I don't think SLD/MapBox should be the default output format.  Of course SLD/MapBox should be the expected *input* format when POSTing or PUTting styles to the .../styles and .../styles/{styleId} endpoints respectively.

To be fair, I don't feel strongly about this.  It's just what makes more sense to me.

The fact that "f=json" can ambiguously mean "OGC API JSON" or "MBGL JSON" is unfortunate.  Could you use "f=mbgl" for the latter instead?

**Comment by pomakis** [comment]
FYI, I've finished implementing these experimental endpoints, including their HTML representations.  Therefore, @jerstlouis's example URL:

https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/styles/Topographic

now has a proper HTML representation, etc.

NOTE: The OpenAPI document still doesn't reflect these new endpoints.  I'm hesitant to put *too* much work into these new endpoints in case for some reason they don't get accepted.

**Comment by jerstlouis** [comment]
@pomakis 
What about collection-specific resources for the style sheet though? i.e. that same end-point but SLD representation:

https://test.cubewerx.com/cubewerx/cubeserv/experimentalMapEndpoints/ogcapi/Daraa/collections/TransportationGroundCrv/styles/Topographic?f=sld10

In our server, those collection-specific end-points have filtered SLD stylesheets with only the rules for that particular collection.

Right now you're pointing to the dataset-wide style sheets.

**Comment by pomakis** [comment]
Oops, yes, that was an oversight.  It's hard to juggle both implementations.  (I'm trying to keep the old endpoints and their behaviours active at https://test.cubewerx.com/cubewerx/cubeserv/demo/ogcapi/Daraa until these new endpoints are more widely approved.)  Fixed.

Our server isn't equipped to PUT/POST/DELETE to the collection-specific style endpoints though, just to the top-level style endpoints.  That could be added in the future though.

**Comment by joanma747** [comment]
We agreed that we will give this a try and try to write the standards (and the examples) this way (in the telecon 2020-10-22)

**Comment by joanma747** [comment]
In the telco today we revisited this issue and tried to classify the proposed path above in conformance classes.

```txt
(essential building block also for not OGC "people")
.../{tilemetrix}/{tilerow}/{tilecol}  defined in OGC API tiles; tileset conformance class (a.k.a. core)

(building block with tileset metadata)
.../tiles                      defined in OGC API tiles; tileset lists conformance class 
.../tiles/{tilematrixsetid}    defined in OGC API tiles; tileset metadata conformance class
.../tiles/{tilematrixsetid}/{tilemetrix}/{tilerow}/{tilecol}  idem 

(how to apply the tiles building block to a {datasetAPI})
{datasetAPI}/collections/{collectionId}/tiles  defined in OGC API tiles; collection
{datasetAPI}/tiles                             defined in OGC API tiles; root (dataset)
{datasetAPI}/tiles?collections=                defined in OGC API tiles; collections

(building block)
.../map                   defined in the OGC API maps; core (default style)
.../map?BBOX=...          defined in the OGC API maps; bbox conformance class (default style)

(how to combine the two building blocks (maps, tiles))
.../map/tiles             defined in the OGC API maps; tiles conformance class (has a default style)

(how to combine the two building blocks and apply them to a {datasetAPI} to get map tiles)
{datasetAPI}/collections/{collectionId}/map/tiles  defined in OGC API tiles; collection
{datasetAPI}/map/tiles                             defined in OGC API tiles; root (dataset, default style)
{datasetAPI}/map/tiles?collections=                defined in OGC API tiles; collections

(how to apply a map building block to a style)
{datasetAPI}/collections/{collectionId}/styles/{styleId}/map         defined in the OGC API maps; styles conformance class
{datasetAPI}/collections/{collectionId}/styles/{styleId}/map/tiles   defined in the OGC API maps; tiles with styles conformance class
{datasetAPI}/styles/{styleId}                                        defined in the OGC API styles  (links to retrieve a style descriptions: SLD...)

(how to combine coverages and tiles building blocks) (out of scope)
.../coverage/tiles             defined in the OGC API maps; tiles conformance class (has a default style)
```

**Comment by jerstlouis** [comment]
See also our previous discussions on this matter:

https://github.com/opengeospatial/ogcapi-maps/issues/32
https://github.com/opengeospatial/ogcapi-tiles/issues/41

The conformance class organization seems quite in line, so that's a very good sign :)

I know we talked about the "not OGC people" conformance class, but maybe that was after this Tiles issue 41 discussion as that is the one thing missing -- though I agree it is a good thing to have that separate.

Regarding the arbitrary vs. prescribed path for different kind of tiles which we discussed towards the end of the meeting, after looking back at what we agreed on last year, I believe the idea is that:
- The Core (just the tiles) conformance class has no fixed path
- The tileset metadata has no fixed path
- The tilesets list has no fixed path.

Clients can therefore follow link from wherever, and if tileset metadata are available they can use that to determine which type of tileset (e.g. map tile / vector tile / coverage tile) it is, what style, etc.

However the challenge is to determine clear mechanics on how the client goes from a dataset landing page to the list of tilesets it cares about. I think having dedicated locations for specific kind of tiles (e.g. dataset tiles vs. collection tiles; tiles for a particular style; vector vs. coverage vs. map tiles) for the _other_ conformance classes (dataset tiles; collection tiles; map tiles; coverage tiles (in OGC API - Coverages) makes a lot of sense.
